File: 백준/Gold/2668. 숫자고르기/숫자고르기.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

bool visited[101], finished[101];
int match[101], result[101];
int idx;

void dfs(int x){
    visited[x] = true;
    int next = match[x];
    if(!visited[next]) dfs(next);
    else if(!finished[next]){
        for(int i = next; i != x; i = match[i]) result[idx++] = i;
        result[idx++] = x;
    }
    finished[x] = true;
}

int compare(const void *a, const void *b){
    int A = *(int*)a;
    int B = *(int*)b;
    if(A > B) return 1;
    return -1;
}

int main(void){
    int N;
    scanf("%d", &N);
    for(int i = 1; i <= N; i++) scanf("%d", &match[i]);
    for(int i = 1; i <= N; i++) if(!visited[i]) dfs(i);
    qsort(result, idx, sizeof(int), compare);

    printf("%d\n", idx);
    for(int i = 0; i < idx; i++) printf("%d\n", result[i]);
}

File: 백준/Gold/2668. 숫자고르기/숫자고르기.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(1000000)
N = int(input())

match = [0]
visited = [False for _ in range(N + 1)]
finished = [False for _ in range(N + 1)]
result = []

def dfs(now):
    global match, visited, finished, result
    visited[now] = True
    next = match[now]

    if not visited[next]: dfs(next)
    elif not finished[next]:
        i = next
        while i != now:
            result.append(i)
            i = match[i]
        result.append(i)
    finished[now] = True

for _ in range(N): match.append(int(input()))
for i in range(1, N + 1):
    if visited[i]: continue
    dfs(i)

result = sorted(result)
cnt = 0
ans = []
for x in result:
    cnt += 1
    ans.append(str(x))
print(cnt)
print('\n'.join(ans))